x-pack/plugin/security/src/main/java/module-info.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+module org.elasticsearch.security {
+    requires java.naming;
+    requires java.security.jgss;
+    requires java.xml;
+
+    requires org.elasticsearch.base;
+    requires org.elasticsearch.cli;
+    requires org.elasticsearch.server;
+    requires org.elasticsearch.ssl.config;
+    requires org.elasticsearch.transport.netty4;
+    requires org.elasticsearch.xcontent;
+    requires org.elasticsearch.xcore;
+
+    requires org.apache.commons.codec;
+    requires org.apache.httpcomponents.httpcore;
+    requires org.apache.httpcomponents.httpclient;
+    requires org.apache.httpcomponents.httpasyncclient;
+    requires org.apache.httpcomponents.httpcore.nio;
+    requires org.apache.log4j;
+    requires org.apache.logging.log4j;
+    requires org.apache.lucene.core;
+    requires org.apache.lucene.queries;
+    requires org.apache.lucene.sandbox;
+
+    requires org.opensaml.core;
+    requires org.opensaml.saml;
+    requires org.opensaml.saml.impl;
+    requires org.opensaml.security.impl;
+    requires org.opensaml.security;
+    requires org.opensaml.xmlsec.impl;
+    requires org.opensaml.xmlsec;
+
+    requires com.nimbusds.jose.jwt;
+    requires io.netty.handler;
+    requires io.netty.transport;
+    requires jopt.simple;
+    requires json.smart;
+    requires net.shibboleth.utilities.java.support;
+    requires oauth2.oidc.sdk;
+    requires slf4j.api;
+    requires unboundid.ldapsdk;
+
+    exports org.elasticsearch.xpack.security.action to org.elasticsearch.server;
```

Comment:
I'm confused by this. I understand this as there being a dependency in server on x-pack security? That can't be the case right as that would introduce a cycle. What is the purpose of these qualified exports?

Reply:
In general, most leaf modules should not need to export anything (or at least not export much). I don't remember why this was needed, I'll figure out why, then report back here.

Reply:
Ha! I remember now. The server uses dependency injection to create instances of these security actions. So server is reflectively accessing classes in these packages. Exports is used here, rather than `opens`, since the server is only using public members - so the deep reflective access of `opens` is not needed.